docs/sources/community/lids/_DRAFT_AsynchronousQueries.md

```diff
@@ -0,0 +1,105 @@
+
+## Problem Statement
+
+Expensive/slow queries can be expressed in Loki terms in many different ways with various effects to the user experience, i.e.:
+
+* A query that has a too-wide label selector might require to scan a tremendous amount of chunks (i.e. effectively TB of logs or more) (e.g. parsing logs across a fleet of clusters).
+* A query that aggregates metrics over a very large time range (i.e. multiple months) and/or chunks (i.e. TB of logs or more) (e.g. count audit log events over months).
+* A query using expensive/slow regex/pattern parsing over a considerable amount of chunks (i.e. TB of logs or more) (e.g. parsing the access log from a high traffic gateway).
+
+Although the list is not exhaustive all examples granted can potentially result either in easily reaching limits (chunks, series, per-request, etc.) or take too long so that client applications and related networking infrastructure run in timeouts. Limits might be considered as tunable but timeouts might not be easy to mitigate or instantly addressable. These use cases can be considered problematic when the requesting user requires:
+
+* The query result without the ability to tune execution time or Loki's resource constraints.
+* The query result represents an expensive but important (aggregation) result set of logs and has a higher value as such irrespective how long it takes.
+
+## Goals
+
+The following proposals aim at providing the required infrastructure and query capabilities to execute queries in Loki asynchronously meaning the user will not fetch a result immediately but at later point in time using an identification for the requested query.
+
+## Non-Goals (optional)
+
+The following proposal do not aim at extending Loki to asynchronous capabilities beyond executing queries, e.g. offering asynchronous streams to read/write from/to Loki.
+
+## Proposals
+
+### Proposal 0: Do nothing
+
+The current state represents the do-nothing scenario very well, as per trade-offs currently to consider:
+
+* Users need to tune their queries manually (e.g. distinct label selectors, batching time frames).
+* Users need to request per-tenant and/or per-request limits by their respecting Loki administrators to overcome limits for certain query demand over time.
+* Client applications and networking infrastructure needs to be tuned for Loki response latency scenarios (e.g. Grafana datasource dataproxy, Server Write Timeouts on Gateways)
+
+### Proposal 1: Built-in asynchronous streaming component for asynchronous queries
```

Comment:
Thanks for putting this together. A few notes:

* I think we'll want to provide storage (likely object storage) + discovery for query requests and status as an API.
* I'm not sure what the added benefit `nats` brings here over object storage + a component to move async queries towards completion (basically a `ruler` that reads requests from object storage and writes responses back to object storage, then combines partial responses into total ones, such as individual steps in a range query).
* We could get fancier and try to optimize running queries when there is low system load, update status based on partial completion (75% -> 80%, etc)

Reply:
> Thanks for putting this together. A few notes:
> 
>     * I think we'll want to provide storage (likely object storage) + discovery for query requests and status as an API.
> 

This sounds like another worthy proposal section in this LID. It sound like a self-baked solution that persists query request and could possible re-use the remote-ruler evaluation like pipeline to execute the queries.

>     * I'm not sure what the added benefit `nats` brings here over object storage + a component to move async queries towards completion (basically a `ruler` that reads requests from object storage and writes responses back to object storage, then combines partial responses into total ones, such as individual steps in a range query).
> 

Although `nats` can store the streams in object storage (currently experimental support but for sure coming), I believe asynchronous request can benefit from the message handling mechanism implemented, i.e.:
- It supports replication and retention out of the box regardless storage on disk or object storage.
- It supports message delivery acks/nacks/etc. so that in case of failure a re-process is an easy piece.

There are a couple of examples and not meant to be comprehensive. I think it is worth having a conversation to evaluate if we need such message queue handling. These usually are there for failure scenarios.

>     * We could get fancier and try to optimize running queries when there is low system load, update status based on partial completion (75% -> 80%, etc)

Sure a status API should be mandatory in the final implementation. The users want to experience that progress is made over time.

Reply:
Fair points. I'm personally leaning away from nats as my gut says the benefits aren't worth the additional complexity here, but open to the idea. Overall I like the idea of natively supporting async queries. If there aren't any detractors, I think the next step would be a more detailed design doc. Nice work 👍